File: cloakbrowser/human/scroll.py
"""cloakbrowser-human — Human-like scrolling via mouse wheel events."""

from __future__ import annotations

import math
import random
import time
from typing import Any, Callable, Optional, Tuple

from .config import HumanConfig, rand, rand_range, rand_int_range, sleep_ms
from .mouse import RawMouse, human_move
from .stealth_dom import (
    build_box_js, eval_parsed, EVALUATION_FAILED, NOT_FOUND, OK, UNSUPPORTED,
    StealthEvaluationError, StealthWorldUnavailableError,
    UnsupportedHumanizeSelectorError, _VIEWPORT_JS,
)


def _is_in_viewport(bounds: dict, viewport_height: int, cfg: HumanConfig) -> bool:
    top_edge = bounds["y"]
    bottom_edge = bounds["y"] + bounds["height"]
    zone_top = viewport_height * cfg.scroll_target_zone[0]
    zone_bottom = viewport_height * cfg.scroll_target_zone[1]
    return top_edge >= zone_top and bottom_edge <= zone_bottom


def _get_element_box(page: Any, selector: str, timeout: float = 30000) -> Optional[dict]:
    """Locate ``selector`` and read geometry only in the isolated world."""
    world = getattr(page, "_stealth_world", None)
    if world is None:
        raise StealthWorldUnavailableError()
    deadline = time.monotonic() + max(0, timeout) / 1000.0
    status, data = eval_parsed(world, build_box_js(selector))
    while status in (NOT_FOUND, EVALUATION_FAILED) and time.monotonic() < deadline:
        time.sleep(0.05)
        status, data = eval_parsed(world, build_box_js(selector))
    if status == OK:
        box = dict(data["box"])
        box["targetId"] = data["targetId"]
        box["gen"] = data["gen"]
        return box
    if status == NOT_FOUND:
        return None
    if status == UNSUPPORTED:
        raise UnsupportedHumanizeSelectorError(selector)
    raise StealthEvaluationError(selector)


_SCROLL_JS = (
    "(() => { const e = document.scrollingElement || document.documentElement;"
    " return { y: window.scrollY, maxY: Math.max(0, e.scrollHeight - e.clientHeight) }; })()"
)


def _read_scroll_state(page: Any) -> dict:
    """Read vertical scroll state only through the isolated world."""
    world = getattr(page, "_stealth_world", None)
    if world is None:
        raise StealthWorldUnavailableError()
    try:
        state = world.evaluate(_SCROLL_JS)
    except Exception as exc:
        raise StealthEvaluationError("<scroll-state>") from exc
    if not isinstance(state, dict):
        raise StealthEvaluationError("<scroll-state>")
    return state


def _smooth_wheel(raw: RawMouse, delta: int, cfg: HumanConfig) -> None:
    """Send one logical scroll as a burst of small wheel events (like real inertia)."""
    abs_d = abs(delta)
    sign = 1 if delta > 0 else -1
    sent = 0
    while sent < abs_d:
        step_size = rand(20, 40)
        chunk = min(step_size, abs_d - sent)
        raw.wheel(0, round(chunk) * sign)
        sent += chunk
        sleep_ms(rand(8, 20))


def human_scroll_into_view(
    page: Any,
    raw: RawMouse,
    get_box: Callable[[], Optional[dict]],
    cursor_x: float, cursor_y: float,
    cfg: HumanConfig,
) -> Tuple[dict, float, float, bool]:
    """Humanized scrolling that uses an arbitrary ``get_box`` callable
    instead of a CSS selector.

    Used both by ``scroll_to_element`` (selector-based) and by
    ``ElementHandle.scroll_into_view_if_needed`` / ``Locator.scroll_into_view_if_needed``
    (handle-based) so the same accelerate \u2192 cruise \u2192 decelerate \u2192 overshoot
    behavior runs everywhere.

    Returns ``(box, cursor_x, cursor_y, did_scroll)`` \u2014 *did_scroll* is False
    when the element was already in the viewport.
    """
    viewport = page.viewport_size
    if not viewport:
        # Headed launches default to no_viewport so the page tracks the real OS
        # window; page.viewport_size is then None. Read the live window dimensions
        # through the isolated world, consistent with the other geometry reads here.
        world = getattr(page, "_stealth_world", None)
        if world is None:
            raise StealthWorldUnavailableError()
        try:
            viewport = world.evaluate(_VIEWPORT_JS)
        except Exception as exc:
            raise StealthEvaluationError("<viewport>") from exc
    if not viewport or not viewport.get("height"):
        raise RuntimeError("Viewport size not available")

    viewport_height = viewport["height"]
    viewport_width = viewport["width"]

    box = get_box()
    if box is None:
        raise RuntimeError("Element not found while scrolling into view")

    if _is_in_viewport(box, viewport_height, cfg):
        return box, cursor_x, cursor_y, False

    # Already fully visible but off-center, with the page pinned at the boundary
    # in the needed direction: scrolling can't help, so don't waste the budget.
    fully_visible = box["y"] >= 0 and box["y"] + box["height"] <= viewport_height
    if fully_visible:
        zone_mid = viewport_height * (cfg.scroll_target_zone[0] + cfg.scroll_target_zone[1]) / 2
        need_up = box["y"] + box["height"] / 2 < zone_mid
        scroll = _read_scroll_state(page)
        if (scroll["y"] <= 0) if need_up else (scroll["y"] >= scroll["maxY"]):
            return box, cursor_x, cursor_y, False

    # Move cursor into scroll area
    scroll_area_x = round(viewport_width * rand(0.3, 0.7))
    scroll_area_y = round(viewport_height * rand(0.3, 0.7))
    human_move(raw, cursor_x, cursor_y, scroll_area_x, scroll_area_y, cfg)
    cursor_x = scroll_area_x
    cursor_y = scroll_area_y
    sleep_ms(rand_range(cfg.scroll_pre_move_delay))

    # Calculate scroll distance
    target_y = viewport_height * rand(cfg.scroll_target_zone[0], cfg.scroll_target_zone[1])
    element_center = box["y"] + box["height"] / 2
    distance_to_scroll = element_center - target_y

    direction = 1 if distance_to_scroll > 0 else -1
    abs_distance = abs(distance_to_scroll)
    avg_delta = (cfg.scroll_delta_base[0] + cfg.scroll_delta_base[1]) / 2
    total_clicks = max(3, math.ceil(abs_distance / avg_delta))
    accel_steps = rand_int_range(cfg.scroll_accel_steps)
    decel_steps = rand_int_range(cfg.scroll_decel_steps)

    # Scroll loop: accelerate → cruise → decelerate
    scrolled = 0
    for i in range(total_clicks):
        if i < accel_steps:
            delta = rand(80, 100)
            pause = rand_range(cfg.scroll_pause_slow)
        elif i >= total_clicks - decel_steps:
            delta = rand(60, 90)
            pause = rand_range(cfg.scroll_pause_slow)
        else:
            delta = rand_range(cfg.scroll_delta_base)
            pause = rand_range(cfg.scroll_pause_fast)

        delta *= 1 + (random.random() - 0.5) * 2 * cfg.scroll_delta_variance
        delta = round(delta) * direction

        _smooth_wheel(raw, delta, cfg)
        scrolled += abs(delta)
        sleep_ms(pause)

        # Check visibility every 3 steps
        if i % 3 == 2 or i == total_clicks - 1:
            box = get_box()
            if box and _is_in_viewport(box, viewport_height, cfg):
                break
        if scrolled >= abs_distance * 1.1:
            break

    # Optional overshoot + correction
    if random.random() < cfg.scroll_overshoot_chance:
        overshoot_px = round(rand_range(cfg.scroll_overshoot_px)) * direction
        _smooth_wheel(raw, overshoot_px, cfg)
        sleep_ms(rand_range(cfg.scroll_settle_delay))
        corrections = rand_int_range((1, 2))
        for _ in range(corrections):
            corr_delta = round(rand(40, 80)) * -direction
            _smooth_wheel(raw, corr_delta, cfg)
            sleep_ms(rand(100, 250))

    # Settle
    sleep_ms(rand_range(cfg.scroll_settle_delay))

    box = get_box()
    if box is None:
        raise RuntimeError("Element lost after scrolling into view")

    return box, cursor_x, cursor_y, True


def scroll_to_element(
    page: Any,
    raw: RawMouse,
    selector: str,
    cursor_x: float, cursor_y: float,
    cfg: HumanConfig,
    timeout: float = 30000,
) -> Tuple[dict, float, float, bool]:
    """Selector-based humanized scroll.

    ``timeout`` bounds isolated-world geometry polling so callers such as
    ``page.click('#x', timeout=5000)`` can wait for slow elements (#172).

    Returns ``(box, cursor_x, cursor_y, did_scroll)``.
    """
    return human_scroll_into_view(
        page, raw,
        lambda: _get_element_box(page, selector, timeout),
        cursor_x, cursor_y, cfg,
    )

File: js/src/human/scroll.ts
/**
 * cloakbrowser-human — Human-like scrolling via mouse wheel events.
 *
 * Selector geometry and live DOM scroll state are read only through the CDP
 * isolated world. ElementHandle callers may still supply their own exact box.
 */

import type { Page } from 'playwright-core';
import { type HumanConfig, rand, randRange, randIntRange, sleep } from './config.js';
import { type RawMouse, humanMove } from './mouse.js';
import {
  buildBoxJs, evalParsed, getWorld, OK, NOT_FOUND, UNSUPPORTED,
  EVALUATION_FAILED, VIEWPORT_JS, StealthEvaluationError,
  StealthWorldUnavailableError, UnsupportedHumanizeSelectorError,
} from './stealthDom.js';

export interface ElementBounds {
  x: number;
  y: number;
  width: number;
  height: number;
  targetId?: number;
}

export interface SelectorBounds extends ElementBounds {
  targetId: number;
  gen: number;
}

function isInViewport(
  bounds: ElementBounds,
  viewportHeight: number,
  cfg: HumanConfig,
): boolean {
  const topEdge = bounds.y;
  const bottomEdge = bounds.y + bounds.height;
  const zoneTop = viewportHeight * cfg.scroll_target_zone[0];
  const zoneBottom = viewportHeight * cfg.scroll_target_zone[1];
  return topEdge >= zoneTop && bottomEdge <= zoneBottom;
}

const SCROLL_JS =
  '(() => { const e = document.scrollingElement || document.documentElement;' +
  ' return { y: window.scrollY, maxY: Math.max(0, e.scrollHeight - e.clientHeight) }; })()';

async function readScrollState(page: Page): Promise<{ y: number; maxY: number }> {
  const world = getWorld(page);
  if (!world) throw new StealthWorldUnavailableError();
  try {
    const state = await world.evaluate(SCROLL_JS);
    if (
      !state || typeof state !== 'object' ||
      typeof state.y !== 'number' || typeof state.maxY !== 'number'
    ) {
      throw new StealthEvaluationError('<scroll-state>');
    }
    return state;
  } catch (error) {
    if (error instanceof StealthEvaluationError) throw error;
    throw new StealthEvaluationError('<scroll-state>');
  }
}

async function smoothWheel(raw: RawMouse, delta: number, cfg: HumanConfig): Promise<void> {
  const absD = Math.abs(delta);
  const sign = delta > 0 ? 1 : -1;
  let sent = 0;
  while (sent < absD) {
    const stepSize = rand(20, 40);
    const chunk = Math.min(stepSize, absD - sent);
    await raw.wheel(0, Math.round(chunk) * sign);
    sent += chunk;
    await sleep(rand(8, 20));
  }
}

export async function humanScrollIntoView<T extends ElementBounds>(
  page: Page,
  raw: RawMouse,
  getBox: () => Promise<T | null>,
  cursorX: number,
  cursorY: number,
  cfg: HumanConfig,
): Promise<{ box: T; cursorX: number; cursorY: number; didScroll: boolean }> {
  let viewport = page.viewportSize();
  if (!viewport) {
    const world = getWorld(page);
    if (!world) throw new StealthWorldUnavailableError();
    try {
      viewport = await world.evaluate(VIEWPORT_JS);
    } catch {
      throw new StealthEvaluationError('<viewport>');
    }
  }
  if (!viewport || !viewport.height) throw new Error('Viewport size not available');

  let box = await getBox();
  if (!box) throw new Error('Element not found while scrolling into view');

  if (isInViewport(box, viewport.height, cfg)) {
    return { box, cursorX, cursorY, didScroll: false };
  }

  const fullyVisible = box.y >= 0 && box.y + box.height <= viewport.height;
  if (fullyVisible) {
    const zoneMid = viewport.height * (cfg.scroll_target_zone[0] + cfg.scroll_target_zone[1]) / 2;
    const needUp = box.y + box.height / 2 < zoneMid;
    const { y, maxY } = await readScrollState(page);
    if (needUp ? y <= 0 : y >= maxY) {
      return { box, cursorX, cursorY, didScroll: false };
    }
  }

  const scrollAreaX = Math.round(viewport.width * rand(0.3, 0.7));
  const scrollAreaY = Math.round(viewport.height * rand(0.3, 0.7));
  await humanMove(raw, cursorX, cursorY, scrollAreaX, scrollAreaY, cfg);
  cursorX = scrollAreaX;
  cursorY = scrollAreaY;
  await sleep(randRange(cfg.scroll_pre_move_delay));

  const targetY = viewport.height * rand(cfg.scroll_target_zone[0], cfg.scroll_target_zone[1]);
  const elementCenter = box.y + box.height / 2;
  const distanceToScroll = elementCenter - targetY;

  const direction = distanceToScroll > 0 ? 1 : -1;
  const absDistance = Math.abs(distanceToScroll);
  const avgDelta = (cfg.scroll_delta_base[0] + cfg.scroll_delta_base[1]) / 2;
  const totalClicks = Math.max(3, Math.ceil(absDistance / avgDelta));
  const accelSteps = randIntRange(cfg.scroll_accel_steps);
  const decelSteps = randIntRange(cfg.scroll_decel_steps);
  let scrolled = 0;

  for (let i = 0; i < totalClicks; i++) {
    let delta: number;
    let pause: number;

    if (i < accelSteps) {
      delta = rand(80, 100);
      pause = randRange(cfg.scroll_pause_slow);
    } else if (i >= totalClicks - decelSteps) {
      delta = rand(60, 90);
      pause = randRange(cfg.scroll_pause_slow);
    } else {
      delta = randRange(cfg.scroll_delta_base);
      pause = randRange(cfg.scroll_pause_fast);
    }

    delta *= 1 + (Math.random() - 0.5) * 2 * cfg.scroll_delta_variance;
    delta = Math.round(delta) * direction;

    await smoothWheel(raw, delta, cfg);
    scrolled += Math.abs(delta);
    await sleep(pause);

    if (i % 3 === 2 || i === totalClicks - 1) {
      const nextBox = await getBox();
      if (nextBox) box = nextBox;
      if (nextBox && isInViewport(nextBox, viewport.height, cfg)) break;
    }
    if (scrolled >= absDistance * 1.1) break;
  }

  if (Math.random() < cfg.scroll_overshoot_chance) {
    const overshootPx = Math.round(randRange(cfg.scroll_overshoot_px)) * direction;
    await smoothWheel(raw, overshootPx, cfg);
    await sleep(randRange(cfg.scroll_settle_delay));

    const corrections = randIntRange([1, 2]);
    for (let c = 0; c < corrections; c++) {
      const corrDelta = Math.round(rand(40, 80)) * -direction;
      await smoothWheel(raw, corrDelta, cfg);
      await sleep(rand(100, 250));
    }
  }

  await sleep(randRange(cfg.scroll_settle_delay));

  const finalBox = await getBox();
  if (!finalBox) throw new Error('Element lost after scrolling into view');
  return { box: finalBox, cursorX, cursorY, didScroll: true };
}

export async function scrollToElement(
  page: Page,
  raw: RawMouse,
  selector: string,
  cursorX: number,
  cursorY: number,
  cfg: HumanConfig,
  timeout: number = 30000,
): Promise<{ box: SelectorBounds; cursorX: number; cursorY: number; didScroll: boolean }> {
  return humanScrollIntoView(
    page,
    raw,
    () => getElementBox(page, selector, timeout),
    cursorX,
    cursorY,
    cfg,
  );
}

export async function getElementBox(
  page: Page,
  selector: string,
  timeout: number = 30000,
): Promise<SelectorBounds | null> {
  const world = getWorld(page);
  if (!world) throw new StealthWorldUnavailableError();

  const deadline = Date.now() + Math.max(0, timeout);
  let result = await evalParsed(world, buildBoxJs(selector));
  while (
    (result.status === NOT_FOUND || result.status === EVALUATION_FAILED) &&
    Date.now() < deadline
  ) {
    await sleep(50);
    result = await evalParsed(world, buildBoxJs(selector));
  }

  const { status, data } = result;
  if (status === OK && data?.box && Number.isInteger(data.targetId) && Number.isInteger(data.gen)) {
    return { ...data.box, targetId: data.targetId, gen: data.gen };
  }
  if (status === NOT_FOUND) return null;
  if (status === UNSUPPORTED) throw new UnsupportedHumanizeSelectorError(selector);
  throw new StealthEvaluationError(selector);
}
